In [1]:
from bs4 import BeautifulSoup
import random
# Requests
import requests
from tqdm import tqdm
import pandas as pd
import numpy as np
import sys
import psycopg2
from psycopg2 import OperationalError, errorcodes, errors
import datetime
sys.path.append("..")

from src.funciones import obtener_url_decat_deproducto, obtener_url_decat_deproducto2, scrapeo

In [4]:
lista_supermercados = ['mercadona', 'carrefour', 'eroski', 'dia', 'hipercor', 'alcampo']
lista_webs = []
for i in lista_supermercados:
    url_supermercado = f'https://super.facua.org/{i}'
    lista_webs.append(url_supermercado)

lista_webs

['https://super.facua.org/mercadona',
 'https://super.facua.org/carrefour',
 'https://super.facua.org/eroski',
 'https://super.facua.org/dia',
 'https://super.facua.org/hipercor',
 'https://super.facua.org/alcampo']

In [5]:
lista_categorias_supermercados = []
for i in lista_webs:
    hrefs = obtener_url_decat_deproducto(i)
    lista_categorias_supermercados.append(hrefs)

Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if


In [6]:
print(len(lista_categorias_supermercados))
print(len(lista_categorias_supermercados[0]))

6
3


In [7]:
lista_categorias_supermercados
# lista_categorias_supermercados

[['https://super.facua.org/mercadona/aceite-de-girasol/',
  'https://super.facua.org/mercadona/aceite-de-oliva/',
  'https://super.facua.org/mercadona/leche/'],
 ['https://super.facua.org/carrefour/aceite-de-girasol/',
  'https://super.facua.org/carrefour/aceite-de-oliva/',
  'https://super.facua.org/carrefour/leche/'],
 ['https://super.facua.org/eroski/aceite-de-girasol/',
  'https://super.facua.org/eroski/aceite-de-oliva/',
  'https://super.facua.org/eroski/leche/'],
 ['https://super.facua.org/dia/aceite-de-girasol/',
  'https://super.facua.org/dia/aceite-de-oliva/',
  'https://super.facua.org/dia/leche/'],
 ['https://super.facua.org/hipercor/aceite-de-girasol/',
  'https://super.facua.org/hipercor/aceite-de-oliva/',
  'https://super.facua.org/hipercor/leche/'],
 ['https://super.facua.org/alcampo/aceite-de-girasol/',
  'https://super.facua.org/alcampo/aceite-de-oliva/',
  'https://super.facua.org/alcampo/leche/']]

In [8]:
lista_hrefs_articulos_totales_supermercados = []
for i in range(len(lista_categorias_supermercados)):
    for href in (lista_categorias_supermercados[i]):
        lista_href_productos_supermercados = obtener_url_decat_deproducto2(href)
        lista_hrefs_articulos_totales_supermercados.append(lista_href_productos_supermercados)


Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if
Hemos hecho el requests, vamos al if


In [9]:
lista_hrefs_articulos_totales_supermercados[0]

['https://super.facua.org/mercadona/aceite-de-girasol/aceite-de-girasol-refinado-02-hacendado-1-l/',
 'https://super.facua.org/mercadona/aceite-de-girasol/aceite-de-girasol-refinado-02-hacendado-5-l/']

In [10]:
df = pd.DataFrame()

In [11]:
# df1 = pd.DataFrame()
df2 = pd.DataFrame()
for i in tqdm(range(len(lista_hrefs_articulos_totales_supermercados))):
    for href in (lista_hrefs_articulos_totales_supermercados[i]):
        df3 = scrapeo(href)
        df3['Supermercado'] = href.split('/')[-4]
        df3['Categoría'] = href.split('/')[-3]
        df3['Artículo'] = href.split('/')[-2]
        df2 = pd.concat([df2,df3])
    

100%|██████████| 18/18 [05:44<00:00, 19.12s/it]


In [12]:
df2.to_csv('../datos/precios_historico_supermercados.csv', index=False)

In [13]:
df2.head()

Día Precio (€) Variación Supermercado          Categoría  \
0  12/07/2024       1,45         =    mercadona  aceite-de-girasol   
1  13/07/2024       1,45         =    mercadona  aceite-de-girasol   
2  14/07/2024       1,45         =    mercadona  aceite-de-girasol   
3  15/07/2024       1,45         =    mercadona  aceite-de-girasol   
4  16/07/2024       1,45         =    mercadona  aceite-de-girasol   

                                      Artículo  
0  aceite-de-girasol-refinado-02-hacendado-1-l  
1  aceite-de-girasol-refinado-02-hacendado-1-l  
2  aceite-de-girasol-refinado-02-hacendado-1-l  
3  aceite-de-girasol-refinado-02-hacendado-1-l  
4  aceite-de-girasol-refinado-02-hacendado-1-l

In [14]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 135867 entries, 0 to 42
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   Día           135867 non-null  object
 1   Precio (€)    135867 non-null  object
 2   Variación     135867 non-null  object
 3   Supermercado  135867 non-null  object
 4   Categoría     135867 non-null  object
 5   Artículo      135867 non-null  object
dtypes: object(6)
memory usage: 7.3+ MB


In [15]:
df2['Precio (€)'] = df2['Precio (€)'].str.replace(',', '.')

In [16]:
df2['Precio (€)'] =  df2['Precio (€)'].astype(float)

In [17]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 135867 entries, 0 to 42
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Día           135867 non-null  object 
 1   Precio (€)    135867 non-null  float64
 2   Variación     135867 non-null  object 
 3   Supermercado  135867 non-null  object 
 4   Categoría     135867 non-null  object 
 5   Artículo      135867 non-null  object 
dtypes: float64(1), object(5)
memory usage: 7.3+ MB


In [18]:
df2['Día'] = pd.to_datetime(df2['Día'], format = "%d/%m/%Y")

In [19]:
df_categoria = df2[['Categoría']].drop_duplicates().reset_index(drop=True).reset_index()

In [20]:
df_categoria.columns= ['Id_Categoria', 'Categoría']

In [22]:
df_supermercado = df2[['Supermercado']].drop_duplicates().reset_index(drop=True).reset_index()

In [23]:
df_supermercado.columns= ['Id_Supermercado', 'Supermercado']

In [24]:
df_mergeo_1 = df2.merge(df_categoria, how='inner', left_on= 'Categoría', right_on= 'Categoría').drop(columns='Categoría')

In [25]:
df_mergeo_1 = df_mergeo_1.merge(df_supermercado, how='inner', left_on='Supermercado', right_on='Supermercado').drop(columns='Supermercado')

In [26]:
df_mergeo_1['Variación'] = df_mergeo_1['Variación'].str.split(' ').str[0].str.replace(',', '.').str.replace('=', '0')


In [27]:
df_mergeo_1['Variación'] = df_mergeo_1['Variación'].astype(float)

In [29]:
df_mergeo_1['Id_Categoria'] = df_mergeo_1['Id_Categoria'].apply(lambda x: x + 1)
df_categoria['Id_Categoria'] = df_categoria['Id_Categoria'].apply(lambda x:x+1)
df_supermercado['Id_Supermercado'] = df_supermercado['Id_Supermercado'].apply(lambda x: x + 1)
df_mergeo_1['Id_Supermercado'] = df_mergeo_1['Id_Supermercado'].apply(lambda x: x + 1)

In [30]:
df_mergeo_1.sample(5)

Día  Precio (€)  Variación  \
123501 2024-10-23        1.11        0.0   
23063  2024-10-19        2.51        0.0   
128861 2024-08-03        3.15        0.0   
74305  2024-10-17       16.90        0.0   
105520 2024-08-16        3.13        0.0   

                                                 Artículo  Id_Categoria  \
123501  central-lechera-asturiana-leche-desnatada-0-ma...             3   
23063          aceite-de-girasol-borgesol-botella-1-litro             1   
128861             mntbelle-leche-fresca-de-cabra-750-ml-             3   
74305   germanor-aceite-de-oliva-virgen-extra-botella-...             2   
105520  carbonell-aceite-de-oliva-spray-especial-planc...             2   

        Id_Supermercado  
123501                6  
23063                 3  
128861                6  
74305                 5  
105520                6

In [31]:
df_mergeo_1.duplicated().sum()

np.int64(1992)

In [32]:
df_mergeo_1 = df_mergeo_1.drop_duplicates()

In [33]:
df_mergeo_1.duplicated().sum()

np.int64(0)

In [53]:
df_mergeo_1.shape

(133875, 6)

In [34]:
df_mergeo_1.to_csv('../datos/precios_mergeo.csv', index=False)

In [35]:
df_categoria.to_csv('../datos/categorias.csv', index=False)
df_supermercado.to_csv('../datos/supermercados.csv', index=False)

Primera Exploración de Datos.

Haciendo una investigacón en los datos, pude darme cuenta de que algunos supermercados tienen productos duplicados entre sus artículos. Esto lo pude demostrar entrando a las webs de estos supermercados para darme cuenta de lo que sucedía. En un primer momento, hice una toma de datos en el día anterior. Lo que me había arrojado una menor cantidad a la actual (135867). Luego de quitar los duplicados me quedo con la cantidad de (133875). 

Al principio, todos los datos fueron recibidos en formato object, pero para una correcta carga de datos hice algunas modificaciones: cambié la columna 'día' a formato datetime, modifiqué el precio y variación a formato float, y las categorías de ID a formato int.

Hice un describe para ver algunas cositas, donde el precio promedio en general no llega a los 10 € (9.68).

Así mismo pude ver que no hay nulos.

In [ ]:
df_mergeo_1.shape

(133875, 6)

In [ ]:
df_mergeo_1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 133875 entries, 0 to 135866
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   Día              133875 non-null  datetime64[ns]
 1   Precio (€)       133875 non-null  float64       
 2   Variación        133875 non-null  float64       
 3   Artículo         133875 non-null  object        
 4   Id_Categoria     133875 non-null  int64         
 5   Id_Supermercado  133875 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(2), object(1)
memory usage: 7.1+ MB


In [ ]:
df_mergeo_1.describe()

Día     Precio (€)      Variación  \
count                         133875  133875.000000  133875.000000   
mean   2024-08-30 12:36:56.551260160       9.688601      -0.003513   
min              2024-06-22 00:00:00       0.500000     -21.330000   
25%              2024-08-04 00:00:00       1.420000       0.000000   
50%              2024-08-30 00:00:00       5.160000       0.000000   
75%              2024-09-25 00:00:00      11.020000       0.000000   
max              2024-10-26 00:00:00     154.320000      20.690000   
std                              NaN      13.619848       0.530699   

        Id_Categoria  Id_Supermercado  
count  133875.000000    133875.000000  
mean        2.472710         4.235772  
min         1.000000         1.000000  
25%         2.000000         3.000000  
50%         3.000000         5.000000  
75%         3.000000         5.000000  
max         3.000000         6.000000  
std         0.584003         1.499421

In [54]:
df_categoria

Id_Categoria          Categoría
0             1  aceite-de-girasol
1             2    aceite-de-oliva
2             3              leche

In [ ]:
df_mergeo_1.groupby('Id_Categoria')['Precio (€)'].mean()

Id_Categoria
1     4.889268
2    18.615759
3     2.617056
Name: Precio (€), dtype: float64

In [ ]:
df_mergeo_1.groupby('Id_Supermercado')['Precio (€)'].mean()

Id_Supermercado
1     6.657574
2    10.488101
3     6.324794
4     6.607959
5    11.082174
6    11.663299
Name: Precio (€), dtype: float64

In [56]:
df_mergeo_1.groupby(['Id_Supermercado', 'Id_Categoria'])['Precio (€)'].min()

Id_Supermercado  Id_Categoria
1                1               1.45
                 2               2.90
                 3               0.77
2                1               1.45
                 2               1.46
                 3               0.72
3                1               1.39
                 2               1.89
                 3               0.73
4                1               1.49
                 2               2.15
                 3               0.69
5                1               1.40
                 2               1.75
                 3               0.64
6                1               1.00
                 2               2.83
                 3               0.50
Name: Precio (€), dtype: float64

In [57]:
df_mergeo_1.groupby(['Id_Supermercado', 'Id_Categoria'])['Precio (€)'].max()

Id_Supermercado  Id_Categoria
1                1                 6.91
                 2                28.80
                 3                 9.36
2                1                40.67
                 2                63.50
                 3                46.30
3                1                11.99
                 2                64.59
                 3                 6.85
4                1                10.99
                 2                39.65
                 3                14.64
5                1                12.79
                 2               154.32
                 3                16.95
6                1                15.88
                 2                62.99
                 3                23.43
Name: Precio (€), dtype: float64